contracts/allocators/NFTXAllocator.sol

```diff
@@ -0,0 +1,254 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.7.5;
+
+import "../libraries/Address.sol";
+import "../libraries/SafeMath.sol";
+import "../libraries/SafeERC20.sol";
+
+import "../interfaces/IERC20.sol";
+import "../interfaces/ITreasury.sol";
+import "../interfaces/INFTXInventoryStaking.sol";
+import "../interfaces/INFTXLPStaking.sol";
+
+import "../interfaces/allocators/IAllocator.sol";
+
+import "../types/FloorAccessControlled.sol";
+
+
+/**
+ * Contract deploys reserves from treasury into NFTX vaults,
+ * earning interest and rewards.
+ */
+
+contract NFTXAllocator is IAllocator, FloorAccessControlled {
+
+    using SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    /**
+     * @notice describes the token used for staking in NFTX.
+     */
+
+    struct stakingTokenData {
+        uint256 vaultId;
+        address rewardToken;
+        bool isLiquidityPool;
+        bool exists;
+    }
+
+    /**
+     * @notice describes the dividend token minted from staking token.
+     */
+
+    struct dividendTokenData {
+        address underlying; // stakingToken
+        address xToken; // dividendToken
+        bool exists;
+    }
+
+    event TreasuryAssetDeployed(address token, uint256 amount, uint256 value);
+    event TreasuryAssetReturned(address token, uint256 amount, uint256 value);
+
+
+    // NFTX Inventory Staking contract
+    INFTXInventoryStaking internal inventoryStaking;
+
+    // NFTX Liquidity Staking contract
+    INFTXLPStaking internal liquidityStaking;
+
+    // Floor Treasury contract
+    ITreasury internal treasury;
+
+    // Corresponding NFTX token vault data for tokens
+    mapping (address => stakingTokenData) public stakingTokenInfo;
+
+    // Corresponding xTokens for tokens
+    mapping (address => dividendTokenData) public dividendTokenInfo;
+
+
+    /**
+     * @notice initialises the construct with no additional logic.
+     */
+
+    constructor (
+        address _authority,
+        address _inventoryStaking,
+        address _liquidityStaking,
+        address _treasury
+    ) FloorAccessControlled(IFloorAuthority(_authority)) {
+        inventoryStaking = INFTXInventoryStaking(_inventoryStaking);
+        liquidityStaking = INFTXLPStaking(_liquidityStaking);
+
+        treasury = ITreasury(_treasury);
+    }
+
+
+    /**
+     * Deprecated in favour of harvestAll(address _token).
+     */
+
+    function harvest(address _token, uint256 _amount) external override {
+        revert("Method is deprecated in favour of harvestAll(address _token)");
+    }
+
+
+    /**
+     * @notice claims rewards from the vault.
+     */
+
+    function harvestAll(address _token) override external onlyPolicy {
```

Comment:
If rewards are earned while the xPUNKWETH token is held in the Allocator then the Allocator will have rewards to claim.

Wondering if we re-instate the old `harvestAll()` and rename this method to `treasuryHarvestAll()` ?